### ISSUE TITLE
Wrong stopping criterion in constrained optimizers

### ISSUE THREAD

**Issue by valentjn** [SGpp/SGpp#51]
In GitLab by @valentjn on Nov 9, 2017, 09:37

This was reported by Lorenzo Tamellini:

```
2) Again on constrained optimization: unless the max number of function
evaluations is reached, the algorithms stop if for X consecutive
subproblems the three following conditions occur simultaneously: the
increment between two consecutive subproblems is small enough, the
equality constraints are verified up to epsilon, the inequality
constraints are verified up to epsilon. However, in the code I see


    xNew = unconstrainedOptimizer.getOptimalPoint();
     [...]
    x = xNew;
    xNew.sub(x); //@@

    if ((xNew.l2Norm() < theta) && (gx.max() < epsilon) && (hx.maxNorm() < epsilon)) {
        // break check   
    }

i.e. after line @@ xNew is 0 so the test ( xNew.l2Norm() < theta ) is
useless.. I guess this is a bug and what you meant was for the update
x=xNew to be done after the break check.
```

In short, the stopping criterion that the increment in $`x`$ should be small is never checked for the constrained optimizers as the increment is accidentally set to zero.

**Comment by valentjn** [comment]
In GitLab by @valentjn on Nov 9, 2017, 09:37

created branch [`46-wrong-stopping-criterion-in-constrained-optimizers`](https://simsgs.informatik.uni-stuttgart.de:8444/SGpp/SGpp/compare/master...46-wrong-stopping-criterion-in-constrained-optimizers)

**Comment by valentjn** [comment]
In GitLab by @valentjn on Nov 9, 2017, 09:56

mentioned in merge request !32

**Comment by valentjn** [comment]
In GitLab by @valentjn on Nov 20, 2017, 14:10

assigned to @valentjn

**Comment by valentjn** [comment]
In GitLab by @valentjn on Nov 28, 2017, 10:59

closed via merge request !32

**Comment by valentjn** [comment]
In GitLab by @valentjn on Nov 28, 2017, 10:59

closed via commit 99e1f23efbd7ad7a57d36b3feee204a9cc887c8d

**Comment by valentjn** [comment]
In GitLab by @valentjn on Dec 19, 2017, 16:09

mentioned in commit 99e1f23efbd7ad7a57d36b3feee204a9cc887c8d
